### PR TITLE
Enhance detection logic for patch variants

### PR DESCRIPTION
**Pull request by rquitales**
### Proposed Changes

This PR improves our Patch resource detection logic by removing the reliance on a hardcoded list. Hardcoded lists are fragile because new Patch resources introduced in future SDK versions, or those generated dynamically by tools like `crd2pulumi`, would **not** be recognized as Patch variants by our provider. This issue is already present with `CustomResourcePatch` resources, as CustomResource kinds will never be included in a hardcoded list.

The updated logic now compares the resource URN and the object's Kind. Since the token within the URN of a patch variant always follows the pattern `kind + "Patch"`, we can use this for comparison. Previously, the logic simply checked if the URN ended in `Patch`, which was inaccurate due to the possibility of a Custom Resource (CR) Kind ending in `Patch` (e.g., "MyResourcePatch"). This would lead to the Pulumi patch variant URN being `MyResourcePatchPatch`.

#### Testing

- Added integration tests for the YAML, Node.js, and Go languages since CustomResources are overlays and not supported in YAML. Existing tests did not ensure that deleted Patch resources only removed the fields applied, and not the whole object. This behavior is now exercised as well.
- Expanded existing unit tests.

Note: I did not expand the test coverage for Python, Java, and .NET at this time, as porting the test programs for these languages requires additional effort. We may revisit and add tests for these languages in the future.

### Related Issues (Optional)

Fixes: #3023

### CODE REVIEW

File: provider/pkg/provider/provider_test.go [hunk 1]
```diff
@@ -155,44 +155,6 @@ func Test_equalNumbers(t *testing.T) {
 	}
 }
 
-func Test_isPatchURN(t *testing.T) {
```
Comment (rquitales) [inline provider/pkg/provider/provider_test.go:158]: This is duplicated in `provider/pkg/kinds/util_test.go`.

File: provider/pkg/kinds/util.go [hunk 2]
```diff
@@ -4,11 +4,12 @@ import (
...
-// IsPatchURN returns true if the URN is for a Patch resource.
-func IsPatchURN(urn resource.URN) bool {
-	urnS := urn.Type().String()
+// IsPatchResource returns true if it is a Patch resource and also ensures
+// that we don't return true for CustomResources that end with "Patch".
+func IsPatchResource(urn resource.URN, kind string) bool {
+	resourceName := urn.Type().Name().String()
 
-	return PatchQualifiedTypes.Has(urnS)
+	return kind+"Patch" == resourceName
```
Comment (rquitales) [inline provider/pkg/kinds/util.go:12]: The crux of the change.

**Review by blampe — APPROVED** [review]
Brilliant!

**Review by EronWright — CHANGES_REQUESTED** [review]
I don't understand how a non-patch resource of kind `MyResourcePatch` would work.

**Comment by rquitales** [comment]
> I don't understand how a non-patch resource of kind `MyResourcePatch` would work.

For a non-patch resource with Kind `MyResourcePatch`, the urn for the non-patch variant would be something like `kubernetes:gv:MyResourcePatch`. The urn for the patch variant will be `kubernetes:gv:MyResourcePatchPatch`. The `kind` for both patch, and non-patch variant is **always** `MyResourcePatch`.

So, we can determine if a resource is the patch variant or not simply by checking that: `kind+"Patch == pulumiResourceName`.

This PR includes both unit and integration tests to showcase this behavior:
Unit test: https://github.com/pulumi/pulumi-kubernetes/blob/d795fadc001b2d55b1ee87de7b8dfe19968cd262/provider/pkg/kinds/util_test.go#L24-L34
Integration test (nodejs example): https://github.com/pulumi/pulumi-kubernetes/blob/d795fadc001b2d55b1ee87de7b8dfe19968cd262/tests/sdk/java/testdata/patch-resources/nodejs/step2/index.ts#L120-L148

**Review by EronWright — APPROVED** [review]

**Pull request merged**